### PR TITLE
CSHARP-640 - Allow Nullable<T> values in UDTs

### PR DESCRIPTION
**Pull request by collinsauve**
For https://datastax-oss.atlassian.net/browse/CSHARP-640

This still needs some tests around it...  investigating where to put them.

### CODE REVIEW

**Comment by jorgebay** [comment]
Nice findings!

About the implementation, what do you think about moving the logic to `ConvertToDbFromUdtFieldValue()`?

**Comment by collinsauve** [comment]
Np, moved it.

**Comment by collinsauve** [comment]
Hey @jorgebay are there any other tests that exercise `Mapper.Insert` with a table using UDTs?   If so can you point me at them so I can use them as an example?

**Comment by jorgebay** [comment]
I think this is not directly related to the mapper (conversion for udts is made by the default `TypeConverter` from the Mapper), so I think the integration tests should go into: [UdtMappingsTests.cs](https://github.com/datastax/csharp-driver/blob/3.4.0/src/Cassandra.IntegrationTests/Core/UdtMappingsTests.cs).

In any case, there are some unit and integration tests for UDTs with the Mapper and Linq: 
- https://github.com/datastax/csharp-driver/blob/3.4.0/src/Cassandra.Tests/Mapping/InsertTests.cs#L134
- https://github.com/datastax/csharp-driver/blob/3.4.0/src/Cassandra.IntegrationTests/Linq/LinqTable/LinqUdtTests.cs
- https://github.com/datastax/csharp-driver/blob/3.4.0/src/Cassandra.IntegrationTests/Mapping/Tests/Fetch.cs#L342

There is a doc on running integration tests locally: https://github.com/datastax/csharp-driver/wiki/Building-and-running-tests
If you don't want to setup the env for running the integration tests locally, feel free to use our CI to check your tests results (just let us know when its ready for review).

**Comment by collinsauve** [comment]
Sorry for the delay, just managed to get some free time today to work on this.

I've changed UdtMappingTests to perform an insert with a `long?` field in the Contact UDT.   If you want a different name for the field let me know, happy to change it (I'm not very creative).

**Comment by collinsauve** [comment]
Test failure is in LoadBalancingPolicyShortTests... probably not related to my changes.  Is it possible this is an intermittent failure @jorgebay  ?

**Comment by jorgebay** [comment]
The test failure is caused by ccm not able to create a test C* cluster, not related to this change 👍

File: src/Cassandra/Mapping/TypeConversion/TypeConverter.cs [hunk 1]
```diff
@@ -96,6 +96,12 @@ protected TypeConverter()
         /// </summary>
         internal object ConvertToDbFromUdtFieldValue(Type valueType, Type dbType, object value)
         {
+            var nullableType = typeof(Nullable<>).MakeGenericType(dbType);
+            if (valueType == nullableType)
+            {
+                return value == null ? null : nullableType.GetTypeInfo().GetProperty("Value").GetValue(value);
```
Comment (jorgebay) [inline src/Cassandra/Mapping/TypeConversion/TypeConverter.cs]: You can use just `return value` for the nullable type regardless of the underlying value.

**Comment by collinsauve** [comment]
@jorgebay thanks for those suggestions, I've made the changes requested.  Can you review again?

**Comment by jorgebay** [comment]
Thanks for the quick turnaround, I'll look at it today.

**Comment by jorgebay** [comment]
Thanks @collinsauve !
I've merge it, I'll add a more tests to cover using udts with nulls as parameters.

**Comment by collinsauve** [comment]
Thanks @jorgebay.

Do you have a timeline for 3.4.1 ?  Just wondering when to expect this fix (bug is blocking our upgrade to 3.4.x).

**Comment by jorgebay** [comment]
We don't have a target date yet...

**Pull request merged**